lib/elixir/lib/application.ex

```diff
@@ -2,61 +2,88 @@ defmodule Application do
-  Each application will then shutdown its top-level supervisor, if one is
-  available, [which then shuts down its children](Supervisor.html#module-start-and-shutdown).
+  ## The application environment
 
-  From Erlang/OTP 19.1, a SIGTERM from the operating system will automatically
-  translate to `System.stop/0`. Erlang/OTP 20 gives user more explicit control
-  over OS signals via the `:os.set_signal/2` function.
+  The key `env` of an application resource file has a list of tuples that map
+  atoms to terms, and its contents are known as the application *environment*.
+  Note that this environment is unrelated to the operating system environment.
+
+  By default, the environment of an application is an empty list. In a Mix
+  project you can set that key in `application/0`:
+
+      def application do
+        [env: [redis_host: "localhost"]]
+      end
+
+  and the generated application resource file is going to have it included.
+
+  The environment is available after loading the application, which is a process
+  explained later:
+
+      Application.load(:APP_NAME)
+      #=> :ok
 
-  ### Application module callback
+      Application.get_env(:APP_NAME, :redis_host)
+      #=> "localhost"
 
-  An application may start and stop a supervision tree when it boots via
-  the application module callback.
+  In Mix projects, the environment of the application and its dependencies can
+  be overridden via the `config/config.exs` file. If you start the application
+  with Mix, that configuration is available at compile time, and at runtime too,
+  but take into account it is not included in the generated application resource
+  file, and it is not available if you start the application without Mix.
 
-  The first step is to pass the module callback in the application definition
-  in the `mix.exs` file:
+  For example, someone using your application can override its `:redis_host`
+  environment variable as follows:
+
+      config :APP_NAME, redis_host: "redis.local"
+
+  The function `put_env/3` allows dynamic configuration of the application
+  environment, but as a rule of thumb each application is responsible for its
+  own environment. Please do not use the functions in this module for directly
+  accessing or modifying the environment of other applications.
+
+  The application environment can be overriden via the `-config` option of
+  `erl`, as well as command-line flags, as we are going to see below.
+
+  ## The application callback module
+
+  The `mod` key of an application resource file configures an application
+  callback module and start argument:
 
       def application do
         [mod: {MyApp, []}]
       end
 
-  Our application now requires the `MyApp` module to provide an application
-  callback. This can be done by invoking `use Application` in that module and
-  defining a `start/2` callback, for example:
+  This key is optional, only needed for applications implemented as a
```

Comment:
I would rephrase this :) 

This sentence seems to suggest that the core of the application is the supervision tree, but that might not be accurate. Maybe use a expression like "applications with a supervision tree" or "applications that have/use a supervision tree"?

Reply:
I feel also so-so about that sentence, but there are several things to consider here.

On one hand, a "library" application may launch processes and supervisors.

The other thing is that I copied this sentence [from the Erlang documentation](http://erlang.org/doc/man/app.html):

> Key mod is necessary for an application implemented as a supervision tree, otherwise the application controller does not know how to start it. mod can be omitted for applications without processes, typically code libraries, for example, STDLIB.

While it seems to be a bit forced, I am not an Erlang programmer, so I guessed that way to phrase it would probably be idiomatic, giving the source.

But if we can find something better...